File: app/services/analytics/engine.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from datetime import date
from typing import Any


@dataclass
class CategoryStat:
    category: str
    total: float
    count: int
    percentage: float


@dataclass
class EntityStat:
    name: str
    sent_total: float
    received_total: float
    net_flow: float
    transaction_count: int


@dataclass
class SpendingAnalytics:
    total_income: float
    total_expenditure: float
    net_cash_flow: float
    savings_rate: float
    avg_daily: float
    avg_weekly: float
    avg_monthly: float
    by_category: list[CategoryStat]
    by_merchant: list[dict]
    by_entity: list[EntityStat]
    by_day: list[dict]
    by_week: list[dict]
    by_month: list[dict]
    recurring_total: float
    subscription_total: float
    bank_charges_total: float
    largest_debit: dict | None
    largest_credit: dict | None


def _parse_date(d: Any) -> date:
    if isinstance(d, date):
        return d
    from datetime import datetime
    if isinstance(d, str):
        return datetime.fromisoformat(d).date()
    return d


class AnalyticsEngine:
    def compute(self, transactions: list[dict]) -> SpendingAnalytics:
        debits = [t for t in transactions if t.get("type") == "debit" and not t.get("is_duplicate")]
        credits = [t for t in transactions if t.get("type") == "credit" and not t.get("is_duplicate")]

        total_income = sum(float(t["amount"]) for t in credits)
        total_expenditure = sum(float(t["amount"]) for t in debits)
        net = total_income - total_expenditure
        savings_rate = (net / total_income * 100) if total_income > 0 else 0.0

        # Category breakdown
        cat_totals: dict[str, float] = defaultdict(float)
        cat_counts: dict[str, int] = defaultdict(int)
        for t in debits:
            cat = t.get("category") or "Unknown"
            cat_totals[cat] += float(t["amount"])
            cat_counts[cat] += 1

        by_category = sorted(
            [
                CategoryStat(
                    category=cat,
                    total=round(total, 2),
                    count=cat_counts[cat],
                    percentage=round((total / total_expenditure * 100) if total_expenditure > 0 else 0, 2),
                )
                for cat, total in cat_totals.items()
            ],
            key=lambda x: x.total, reverse=True,
        )

        # Merchant breakdown
        merchant_totals: dict[str, dict] = defaultdict(lambda: {"total": 0.0, "count": 0})
        for t in debits:
            m = t.get("merchant")
            if m:
                merchant_totals[m]["total"] += float(t["amount"])
                merchant_totals[m]["count"] += 1

        by_merchant = sorted(
            [{"merchant": k, "total": round(v["total"], 2), "count": v["count"]}
             for k, v in merchant_totals.items()],
            key=lambda x: x["total"], reverse=True,
        )[:15]

        # Entity (people) breakdown
        entity_map: dict[str, dict] = defaultdict(lambda: {"sent": 0.0, "received": 0.0, "count": 0})
        for t in transactions:
            eid = t.get("entity_name") or t.get("merchant")
            if not eid or t.get("category") != "People":
                continue
            if t.get("type") == "debit":
                entity_map[eid]["sent"] += float(t["amount"])
            else:
                entity_map[eid]["received"] += float(t["amount"])
            entity_map[eid]["count"] += 1

        by_entity = sorted(
            [
                EntityStat(
                    name=name,
                    sent_total=round(v["sent"], 2),
                    received_total=round(v["received"], 2),
                    net_flow=round(v["received"] - v["sent"], 2),
                    transaction_count=v["count"],
                )
                for name, v in entity_map.items()
            ],
            key=lambda x: x.sent_total + x.received_total, reverse=True,
        )

        # Time breakdown
        by_day: dict[str, float] = defaultdict(float)
        by_week: dict[str, float] = defaultdict(float)
        by_month: dict[str, dict] = defaultdict(lambda: {"income": 0.0, "expenditure": 0.0})

        for t in transactions:
            d = _parse_date(t["date"])
            month_key = d.strftime("%b %Y")
            if t.get("type") == "debit":
                by_day[str(d)] += float(t["amount"])
                week_key = f"{d.year}-W{d.isocalendar().week:02d}"
                by_week[week_key] += float(t["amount"])
                by_month[month_key]["expenditure"] += float(t["amount"])
            else:
                by_month[month_key]["income"] += float(t["amount"])

        days = len(by_day) or 1
        avg_daily = total_expenditure / days
        avg_weekly = avg_daily * 7
        avg_monthly = avg_daily * 30

        recurring_total = sum(float(t["amount"]) for t in debits if t.get("is_recurring"))
        subscription_total = sum(float(t["amount"]) for t in debits if t.get("is_subscription"))
        bank_charges_total = sum(float(t["amount"]) for t in debits if t.get("is_bank_charge"))

        largest_debit = max(debits, key=lambda t: float(t["amount"]), default=None)
        largest_credit = max(credits, key=lambda t: float(t["amount"]), default=None)

        return SpendingAnalytics(
            total_income=round(total_income, 2),
            total_expenditure=round(total_expenditure, 2),
            net_cash_flow=round(net, 2),
            savings_rate=round(savings_rate, 2),
            avg_daily=round(avg_daily, 2),
            avg_weekly=round(avg_weekly, 2),
            avg_monthly=round(avg_monthly, 2),
            by_category=by_category,
            by_merchant=by_merchant,
            by_entity=by_entity,
            by_day=[{"date": k, "amount": round(v, 2)} for k, v in sorted(by_day.items())],
            by_week=[{"week": k, "amount": round(v, 2)} for k, v in sorted(by_week.items())],
            by_month=[{"month": k, "income": round(v["income"], 2), "expenditure": round(v["expenditure"], 2)}
                      for k, v in sorted(by_month.items())],
            recurring_total=round(recurring_total, 2),
            subscription_total=round(subscription_total, 2),
            bank_charges_total=round(bank_charges_total, 2),
            largest_debit=largest_debit,
            largest_credit=largest_credit,
        )

    def detect_money_leaks(self, current: list[dict], previous: list[dict]) -> list[dict]:
        def cat_totals(txs: list[dict]) -> dict[str, float]:
            totals: dict[str, float] = defaultdict(float)
            for t in txs:
                if t.get("type") == "debit" and not t.get("is_duplicate"):
                    cat = t.get("category") or "Unknown"
                    totals[cat] += float(t["amount"])
            return totals

        curr = cat_totals(current)
        prev = cat_totals(previous)
        leaks = []

        for cat, curr_amt in curr.items():
            prev_amt = prev.get(cat, 0)
            if prev_amt <= 0:
                continue
            change_pct = ((curr_amt - prev_amt) / prev_amt) * 100
            if change_pct >= 15:
                saving = curr_amt * 0.25
                leaks.append({
                    "category": cat,
                    "current_amount": round(curr_amt, 2),
                    "previous_amount": round(prev_amt, 2),
                    "change_percent": round(change_pct, 2),
                    "potential_saving_monthly": round(saving, 2),
                    "potential_saving_annual": round(saving * 12, 2),
                })

        return sorted(leaks, key=lambda x: x["change_percent"], reverse=True)

    def compare_periods(self, current: list[dict], previous: list[dict]) -> dict:
        curr_analytics = self.compute(current)
        prev_analytics = self.compute(previous)

        income_change = ((curr_analytics.total_income - prev_analytics.total_income) / prev_analytics.total_income * 100) if prev_analytics.total_income else 0
        spend_change = ((curr_analytics.total_expenditure - prev_analytics.total_expenditure) / prev_analytics.total_expenditure * 100) if prev_analytics.total_expenditure else 0

        return {
            "current": {
                "income": curr_analytics.total_income,
                "expenditure": curr_analytics.total_expenditure,
                "savings_rate": curr_analytics.savings_rate,
            },
            "previous": {
                "income": prev_analytics.total_income,
                "expenditure": prev_analytics.total_expenditure,
                "savings_rate": prev_analytics.savings_rate,
            },
            "income_change_pct": round(income_change, 2),
            "spend_change_pct": round(spend_change, 2),
        }


analytics_engine = AnalyticsEngine()

File: src/services/analytics/engine.ts
import type {
  SpendingAnalytics,
  CategoryBreakdown,
  MerchantBreakdown,
  EntityBreakdown,
  DayBreakdown,
  PeriodBreakdown,
  NormalizedTransaction,
} from "../../types";

// ─── Analytics Engine — all calculations are deterministic ────────────────────

export class AnalyticsEngine {
  compute(transactions: NormalizedTransaction[]): SpendingAnalytics {
    const debits = transactions.filter(
      (t) => t.type === "debit" && !t.isDuplicate
    );
    const credits = transactions.filter(
      (t) => t.type === "credit" && !t.isDuplicate
    );

    const totalIncome = credits.reduce((s, t) => s + t.amount, 0);
    const totalExpenditure = debits.reduce((s, t) => s + t.amount, 0);
    const netCashFlow = totalIncome - totalExpenditure;
    const savingsRate =
      totalIncome > 0 ? ((totalIncome - totalExpenditure) / totalIncome) * 100 : 0;

    const allAmounts = [...debits, ...credits].sort(
      (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()
    );

    const uniqueDays = this.getUniqueDays(debits);
    const avgDailySpending =
      uniqueDays > 0 ? totalExpenditure / uniqueDays : 0;
    const avgWeeklySpending = avgDailySpending * 7;
    const avgMonthlySpending = avgDailySpending * 30.44;

    const largestTransaction = Math.max(...allAmounts.map((t) => t.amount), 0);
    const transactionCount = transactions.filter((t) => !t.isDuplicate).length;

    const bankChargesTotal = debits
      .filter((t) => t.isBankCharge)
      .reduce((s, t) => s + t.amount, 0);
    const subscriptionTotal = debits
      .filter((t) => t.isSubscription)
      .reduce((s, t) => s + t.amount, 0);
    const recurringTotal = debits
      .filter((t) => t.isRecurring || t.isSubscription)
      .reduce((s, t) => s + t.amount, 0);

    return {
      totalIncome,
      totalExpenditure,
      netCashFlow,
      savingsRate: Math.round(savingsRate * 100) / 100,
      avgDailySpending,
      avgWeeklySpending,
      avgMonthlySpending,
      largestTransaction,
      transactionCount,
      byCategory: this.computeCategoryBreakdown(debits, totalExpenditure),
      byMerchant: this.computeMerchantBreakdown(debits, totalExpenditure),
      byEntity: this.computeEntityBreakdown(transactions),
      byDay: this.computeDayBreakdown(transactions),
      byWeek: this.computePeriodBreakdown(transactions, "week"),
      byMonth: this.computePeriodBreakdown(transactions, "month"),
      recurringTotal,
      subscriptionTotal,
      bankChargesTotal,
    };
  }

  private computeCategoryBreakdown(
    debits: NormalizedTransaction[],
    total: number
  ): CategoryBreakdown[] {
    const map = new Map<string, { amount: number; count: number; subcategories: Map<string, number> }>();

    for (const t of debits) {
      const cat = t.category || "Other";
      const existing = map.get(cat) || {
        amount: 0,
        count: 0,
        subcategories: new Map(),
      };
      existing.amount += t.amount;
      existing.count += 1;
      if (t.subcategory) {
        existing.subcategories.set(
          t.subcategory,
          (existing.subcategories.get(t.subcategory) || 0) + t.amount
        );
      }
      map.set(cat, existing);
    }

    const result: CategoryBreakdown[] = [];
    for (const [cat, data] of map) {
      result.push({
        category: cat,
        amount: data.amount,
        count: data.count,
        percentage: total > 0 ? (data.amount / total) * 100 : 0,
      });
    }

    return result.sort((a, b) => b.amount - a.amount);
  }

  private computeMerchantBreakdown(
    debits: NormalizedTransaction[],
    total: number
  ): MerchantBreakdown[] {
    const map = new Map<string, { amount: number; count: number; category?: string }>();

    for (const t of debits) {
      if (!t.merchant) continue;
      const existing = map.get(t.merchant) || { amount: 0, count: 0, category: t.category };
      existing.amount += t.amount;
      existing.count += 1;
      map.set(t.merchant, existing);
    }

    const result: MerchantBreakdown[] = [];
    for (const [merchant, data] of map) {
      result.push({
        merchant,
        amount: data.amount,
        count: data.count,
        percentage: total > 0 ? (data.amount / total) * 100 : 0,
        category: data.category,
      });
    }

    return result.sort((a, b) => b.amount - a.amount).slice(0, 20);
  }

  private computeEntityBreakdown(
    transactions: NormalizedTransaction[]
  ): EntityBreakdown[] {
    const map = new Map<
      string,
      {
        id?: string;
        type?: string;
        sent: number;
        received: number;
        count: number;
        amounts: number[];
      }
    >();

    for (const t of transactions) {
      if (!t.entity) continue;
      const key = t.entity;
      const existing = map.get(key) || {
        sent: 0,
        received: 0,
        count: 0,
        amounts: [],
      };
      if (t.type === "debit") existing.sent += t.amount;
      if (t.type === "credit") existing.received += t.amount;
      existing.count += 1;
      existing.amounts.push(t.amount);
      map.set(key, existing);
    }

    const result: EntityBreakdown[] = [];
    for (const [entity, data] of map) {
      result.push({
        entity,
        totalSent: data.sent,
        totalReceived: data.received,
        netAmount: data.received - data.sent,
        count: data.count,
        averageAmount: data.amounts.reduce((s, a) => s + a, 0) / data.amounts.length,
        largestTransaction: Math.max(...data.amounts),
      });
    }

    return result.sort((a, b) => b.totalSent - a.totalSent);
  }

  private computeDayBreakdown(transactions: NormalizedTransaction[]): DayBreakdown[] {
    const map = new Map<string, { income: number; expenditure: number; count: number }>();

    for (const t of transactions) {
      const dayKey = new Date(t.date).toISOString().slice(0, 10);
      const existing = map.get(dayKey) || { income: 0, expenditure: 0, count: 0 };
      if (t.type === "credit") existing.income += t.amount;
      if (t.type === "debit") existing.expenditure += t.amount;
      existing.count += 1;
      map.set(dayKey, existing);
    }

    return Array.from(map.entries())
      .map(([date, data]) => ({ date, ...data }))
      .sort((a, b) => a.date.localeCompare(b.date));
  }

  private computePeriodBreakdown(
    transactions: NormalizedTransaction[],
    period: "week" | "month"
  ): PeriodBreakdown[] {
    const map = new Map<string, { income: number; expenditure: number; count: number }>();

    for (const t of transactions) {
      const d = new Date(t.date);
      let key: string;
      if (period === "month") {
        key = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`;
      } else {
        // ISO week
        const thursday = new Date(d.getTime());
        thursday.setDate(d.getDate() - ((d.getDay() + 6) % 7) + 3);
        const jan4 = new Date(thursday.getFullYear(), 0, 4);
        const week = Math.ceil(
          ((thursday.getTime() - jan4.getTime()) / 86400000 + 1) / 7
        );
        key = `${thursday.getFullYear()}-W${String(week).padStart(2, "0")}`;
      }

      const existing = map.get(key) || { income: 0, expenditure: 0, count: 0 };
      if (t.type === "credit") existing.income += t.amount;
      if (t.type === "debit") existing.expenditure += t.amount;
      existing.count += 1;
      map.set(key, existing);
    }

    return Array.from(map.entries())
      .map(([p, data]) => ({
        period: p,
        income: data.income,
        expenditure: data.expenditure,
        savings: data.income - data.expenditure,
        count: data.count,
      }))
      .sort((a, b) => a.period.localeCompare(b.period));
  }

  private getUniqueDays(transactions: NormalizedTransaction[]): number {
    const days = new Set(
      transactions.map((t) => new Date(t.date).toISOString().slice(0, 10))
    );
    return days.size || 1;
  }

  // ─── Period Comparison ──────────────────────────────────────────────────────

  compareMonths(
    current: NormalizedTransaction[],
    previous: NormalizedTransaction[]
  ): Record<string, number> {
    const computeTotal = (txs: NormalizedTransaction[], type: string) =>
      txs.filter((t) => t.type === type).reduce((s, t) => s + t.amount, 0);

    const currentSpend = computeTotal(current, "debit");
    const previousSpend = computeTotal(previous, "debit");
    const currentIncome = computeTotal(current, "credit");
    const previousIncome = computeTotal(previous, "credit");

    const spendChange =
      previousSpend > 0
        ? ((currentSpend - previousSpend) / previousSpend) * 100
        : 0;

    const incomeChange =
      previousIncome > 0
        ? ((currentIncome - previousIncome) / previousIncome) * 100
        : 0;

    return {
      currentSpend,
      previousSpend,
      spendChange: Math.round(spendChange * 100) / 100,
      currentIncome,
      previousIncome,
      incomeChange: Math.round(incomeChange * 100) / 100,
    };
  }

  // ─── Money Leak Detection ───────────────────────────────────────────────────

  detectMoneyLeaks(
    current: NormalizedTransaction[],
    previous: NormalizedTransaction[]
  ): Array<{
    category: string;
    currentAmount: number;
    previousAmount: number;
    changePercent: number;
    potentialSavingMonthly: number;
    potentialSavingAnnual: number;
  }> {
    const currentByCategory = this.groupByCategory(current);
    const previousByCategory = this.groupByCategory(previous);
    const leaks = [];

    for (const [cat, currentAmount] of currentByCategory) {
      if (cat === "Income") continue;

      const previousAmount = previousByCategory.get(cat) || 0;
      if (previousAmount === 0 || currentAmount < 1000) continue;

      const changePercent =
        ((currentAmount - previousAmount) / previousAmount) * 100;

      if (changePercent >= 15) {
        const suggestion = 0.25; // Suggest 25% reduction
        const potentialSavingMonthly = currentAmount * suggestion;
        const potentialSavingAnnual = potentialSavingMonthly * 12;

        leaks.push({
          category: cat,
          currentAmount,
          previousAmount,
          changePercent: Math.round(changePercent * 10) / 10,
          potentialSavingMonthly: Math.round(potentialSavingMonthly),
          potentialSavingAnnual: Math.round(potentialSavingAnnual),
        });
      }
    }

    return leaks.sort((a, b) => b.changePercent - a.changePercent);
  }

  private groupByCategory(
    transactions: NormalizedTransaction[]
  ): Map<string, number> {
    const map = new Map<string, number>();
    for (const t of transactions.filter((t) => t.type === "debit")) {
      const cat = t.category || "Other";
      map.set(cat, (map.get(cat) || 0) + t.amount);
    }
    return map;
  }
}

export const analyticsEngine = new AnalyticsEngine();
